tests/handlers/test_user_directory.py

```diff
@@ -639,6 +640,64 @@ def _add_user_to_room(
             self.hs.get_storage().persistence.persist_event(event, context)
         )
 
+    def test_making_room_public_doesnt_alter_directory_entry(self):
+        """Per-room names shouldn't go to the directory when the room becomes public.
+
+        I made this a synapse test case rather than a complement one because
+        I think this is (strictly speaking) an implementation choice. Synapse
+        has chosen to only ever use the public profile when responding to a user
+        directory search. There's no privacy leak here, because making the room
+        public discloses the per-room name.
+
+        The spec doesn't mandate anything about _how_ a user
+        should appear in a /user_directory/search result. Hypothetical example:
+        suppose Bob searches for Alice. When representing Alice in a search
+        result, it's reasonable to use any of Alice's nicknames that Bob is
+        aware of. Heck, maybe we even want to use lots of them in a combined
+        displayname like `Alice (aka "ali", "ally", "41iC3")`.
+        """
+        # TODO the same should apply when Alice is a remote user.
+        alice = self.register_user("alice", "pass")
+        alice_token = self.login(alice, "pass")
+        bob = self.register_user("bob", "pass")
+        bob_token = self.login(bob, "pass")
+
+        # Alice and Bob are in a private room.
+        room = self.helper.create_room_as(alice, is_public=False, tok=alice_token)
+        self.helper.invite(room, src=alice, targ=bob, tok=alice_token)
+        self.helper.join(room, user=bob, tok=bob_token)
+
+        # Alice has a nickname unique to that room.
+        self.helper.send_state(
+            room,
+            "m.room.member",
+            {
+                "displayname": "Freddy Mercury",
+                "membership": "join",
+            },
+            alice_token,
+            state_key=alice,
+        )
+
+        # One of them makes the room public.
+        self.helper.send_state(
+            room,
+            "m.room.join_rules",
+            {"join_rule": "public"},
+            alice_token,
+        )
+        # Wait for the handler to process the event
+        self.assertIn((alice, room), self.get_users_in_public_rooms())
+
+        # Alice's nickname is now publicly visible, but this shouldn't change
+        # her entry in the user directory.
```

Comment:
this comment is likely to be confusing when all the context has left us.

I don't think it really needs much other than just remarking it stayed the same.

```suggestion
        # Alice's display name remains the same in the user directory.
```